crates/database/src/order_execution.rs

```diff
@@ -1,33 +1,103 @@
 use {
-    crate::{auction::AuctionId, OrderUid},
+    crate::{auction::AuctionId, Address, OrderUid},
     bigdecimal::BigDecimal,
-    sqlx::PgConnection,
+    sqlx::{PgConnection, QueryBuilder},
+    std::collections::HashMap,
 };
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FeeAsset {
+    pub amount: BigDecimal,
+    pub token: Address,
+}
+
 pub async fn save(
     ex: &mut PgConnection,
     order: &OrderUid,
     auction: AuctionId,
     block_number: i64,
     executed_fee: &BigDecimal,
+    executed_protocol_fees: &[FeeAsset],
 ) -> Result<(), sqlx::Error> {
+    let (protocol_fee_tokens, protocol_fee_amounts) = executed_protocol_fees.iter().fold(
+        (Vec::new(), Vec::new()),
+        |(mut tokens, mut amounts), fee| {
+            tokens.push(fee.token);
+            amounts.push(fee.amount.clone());
+            (tokens, amounts)
+        },
+    );
+
     const QUERY: &str = r#"
-INSERT INTO order_execution (order_uid, auction_id, reward, surplus_fee, block_number)
-VALUES ($1, $2, $3, $4, $5)
+INSERT INTO order_execution (order_uid, auction_id, reward, surplus_fee, block_number, protocol_fee_tokens, protocol_fee_amounts)
+VALUES ($1, $2, $3, $4, $5, $6, $7)
 ON CONFLICT (order_uid, auction_id)
-DO UPDATE SET reward = $3, surplus_fee = $4, block_number = $5
+DO UPDATE SET reward = $3, surplus_fee = $4, block_number = $5, protocol_fee_tokens = $6, protocol_fee_amounts = $7
 ;"#;
     sqlx::query(QUERY)
         .bind(order)
         .bind(auction)
         .bind(0.) // reward is deprecated but saved for historical analysis
         .bind(Some(executed_fee))
         .bind(block_number)
+        .bind(protocol_fee_tokens)
+        .bind(protocol_fee_amounts)
         .execute(ex)
         .await?;
     Ok(())
 }
 
+// fetch protocol fees for all keys in the filter
+pub async fn executed_protocol_fees(
+    ex: &mut PgConnection,
+    keys_filter: &[(AuctionId, OrderUid)],
+) -> Result<HashMap<(AuctionId, OrderUid), Vec<FeeAsset>>, sqlx::Error> {
+    if keys_filter.is_empty() {
+        return Ok(HashMap::new());
+    }
+
+    let mut query_builder = QueryBuilder::new(
+        "SELECT order_uid, auction_id, protocol_fee_tokens, protocol_fee_amounts FROM \
+         order_execution WHERE ",
+    );
+
+    for (i, (auction_id, order_uid)) in keys_filter.iter().enumerate() {
+        if i > 0 {
+            query_builder.push(" OR ");
+        }
+        query_builder
+            .push("(order_uid = ")
+            .push_bind(order_uid)
+            .push(" AND auction_id = ")
+            .push_bind(auction_id)
+            .push(")");
+    }
+
+    #[derive(Clone, Debug, Eq, PartialEq, sqlx::Type, sqlx::FromRow)]
+    struct ProtocolFees {
+        pub order_uid: OrderUid,
+        pub auction_id: AuctionId,
+        pub protocol_fee_tokens: Vec<Address>,
+        pub protocol_fee_amounts: Vec<BigDecimal>,
```

Comment:
is it guarantee in the query that we are retrieving the token and its corresponding amount **in the same order**? it could be dangerous in case the first token doesn't correspond with the first amount, etc

Reply:
Caller is responsible for storing the values in proper order.